compressors/typical_set_coder.py

```diff
@@ -0,0 +1,208 @@
+        num_overall_seen += 1
+
+    return index_typical, index_overall
+
+
+class TypicalSetEncoder(DataEncoder):
+    """Typical set encoder working on blocks of length n. See module documentation for more details."""
+
+    def __init__(self, prob_dist, n, eps) -> None:
+        super().__init__()
+        self.n = n
+        self.eps = eps
+        self.index_typical, self.index_overall = generate_indexes(prob_dist, n, eps)
+        # compute lengths of the encoding of the index for typical and non-typical cases
+        if len(self.index_typical) == 0:
+            self.index_bitlen_typical = None  # empty typical set, so can never encounter
+        else:
+            self.index_bitlen_typical = int(np.ceil(np.log2(len(self.index_typical))))
+        self.index_bitlen_overall = int(np.ceil(np.log2(len(self.index_overall))))
+
+    def encode_block(self, data_block: DataBlock) -> BitArray:
+        encoded_bitarray = BitArray("")
+        # make sure length is exact multiple of block size n
+        assert (
+            data_block.size % self.n == 0
+        ), "Input to typical set encoder should be multiple of block length n"
+        for i in range(0, data_block.size // self.n):
+            block_to_encode = tuple(data_block.data_list[i * self.n : (i + 1) * self.n])
+            if block_to_encode in self.index_typical:
+                # typical case
+                encoded_bitarray += "0"
+                if self.index_bitlen_typical > 0:
+                    encoded_bitarray += uint_to_bitarray(
+                        self.index_typical[block_to_encode], self.index_bitlen_typical
+                    )
+            else:
+                # non-typical case
+                encoded_bitarray += "1"
+                if self.index_bitlen_overall > 0:
+                    encoded_bitarray += uint_to_bitarray(
+                        self.index_overall[block_to_encode], self.index_bitlen_overall
+                    )
+        return encoded_bitarray
+
+
+class TypicalSetDecoder(DataDecoder):
+    """Typical set decoder. See module documentation for more details."""
+
+    def __init__(self, prob_dist, n, eps):
+        super().__init__()
+        self.n = n
+        self.eps = eps
+        self.index_typical, self.index_overall = generate_indexes(prob_dist, n, eps)
+        # compute lengths of the encoding of the index for typical and non-typical cases
+        if len(self.index_typical) == 0:
+            self.index_bitlen_typical = None  # never going to encounter
+        else:
+            self.index_bitlen_typical = int(np.ceil(np.log2(len(self.index_typical))))
+        self.index_bitlen_overall = int(np.ceil(np.log2(len(self.index_overall))))
+
+        # compute inverse index for decoding
+        self.inverse_index_typical = {self.index_typical[s]: s for s in self.index_typical}
+        self.inverse_index_overall = {self.index_overall[s]: s for s in self.index_overall}
+
+    def decode_block(self, bitarray: BitArray):
```

Comment:
might be good to rename `bitarray` to be `encoded_bitarray` for consistency with other compressors

Reply:
I think we didnt do this early on.. I will add a Issue for that

Reply:
done